### PR TITLE
[feat]: [CDS-78369]: Addition of env step to custom stage

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
Added an environment step also to the custom stage plan creator. 
Screenshot 2023-09-12 at 6 14 23 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52719)

### CODE REVIEW

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 163 at r1](https://reviewable.io/reviews/harness/harness-core/52719#-Ne8UbR42UtNyD0H70eh:-Ne8UbR5-DBt3j33JGzX:b-8o9h0d) ([raw file](https://github.com/harness/harness-core/blob/f00b1225777fc1608e464ae2cdd8833e2b49e8f5/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L163)):*
> ```Java
>       final PlanNode envNode =
>           EnvironmentPlanCreatorHelper.getPlanNodeForCustomStage(envNodeUuid, stepParameters, advisorParameters);
>       planCreationResponseMap.put(envNode.getUuid(), PlanCreationResponse.builder().planNode(envNode).build());
> ```

can we extract out this logic into a new method. ?

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 11 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52719)* status: 12 file(s) to review. 0 of 2 approvals obtained.

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/CustomStageConfig.java [hunk 1]
```diff
@@ -44,6 +45,8 @@ public class CustomStageConfig implements StageInfoConfig {
   String uuid;
   @NotNull @VariableExpression(skipVariableExpression = true) private ExecutionElementConfig execution;
 
+  @VariableExpression(skipVariableExpression = true) private EnvironmentYamlV2 environment;
```
Comment (naidu-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/CustomStageConfig.java:48]: this should not be part of custom stage

**Review by naidu-harness — CHANGES_REQUESTED** [review]

**Review by ivan-83 — CHANGES_REQUESTED** [review]
LGTM with minor comments. We need to add and an infrastructure step in the next ticket. When the pipeline starts the Infrastructure step will be executed in the Custom Stage before the other steps. The Infrastructure step will produce environment and infrastructure outcomes. 

> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52719#-:-Ne95ghqDy38rVtXpMqG:b8wy4yq)* status: 1 file(s) to review. 2 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 150 at r1](https://reviewable.io/reviews/harness/harness-core/52719#-Ne93VInA273Pnkh9yzL:-Ne93VIo9bvDqQpIGDR1:bv37jkw) ([raw file](https://github.com/harness/harness-core/blob/f00b1225777fc1608e464ae2cdd8833e2b49e8f5/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L150)):*
> ```Java
> 
>     // Adding Spec node
>     String specNextNodeUuid = envNodeExists ? envNodeUuid : executionFieldUuid;
> ```

I don't get this. We should have both env and execution nodes, or only execution node within the custom stage spec. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 155 at r1](https://reviewable.io/reviews/harness/harness-core/52719#-Ne9-dp42R_x7ramSgql:-Ne9-dp42R_x7ramSgqm:bmgherj) ([raw file](https://github.com/harness/harness-core/blob/f00b1225777fc1608e464ae2cdd8833e2b49e8f5/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L155)):*
> ```Java
> 
>     // Adding Env node
>     if (envNodeExists) {
> ```

this is good, makes the env step optional. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 163 at r1](https://reviewable.io/reviews/harness/harness-core/52719#-Ne8UbR42UtNyD0H70eh:-Ne940_C8bLrHJb631MG:b-t77xdl) ([raw file](https://github.com/harness/harness-core/blob/f00b1225777fc1608e464ae2cdd8833e2b49e8f5/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L163)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we extract out this logic into a new method. ?
</blockquote></details>

agree.

**Review by lovish1999 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52719)* status: 1 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 150 at r1](https://reviewable.io/reviews/harness/harness-core/52719#-Ne93VInA273Pnkh9yzL:-Ne99-LS1ARi1xRk80z0:b-bfw1do) ([raw file](https://github.com/harness/harness-core/blob/f00b1225777fc1608e464ae2cdd8833e2b49e8f5/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L150)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

I don't get this. We should have both env and execution nodes, or only execution node within the custom stage spec.
</blockquote></details>

This basically is to decide which step should be executed after custom stage. So if env is null, then execution steps will start otherwise env. step will be start. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 155 at r1](https://reviewable.io/reviews/harness/harness-core/52719#-Ne9-dp42R_x7ramSgql:-Ne99EWz-YtGzE_S_G4z:b-7g7nkc) ([raw file](https://github.com/harness/harness-core/blob/f00b1225777fc1608e464ae2cdd8833e2b49e8f5/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L155)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

this is good, makes the env step optional.
</blockquote></details>

yes. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 163 at r1](https://reviewable.io/reviews/harness/harness-core/52719#-Ne8UbR42UtNyD0H70eh:-Ne99GBvEOYeIDgf6iPp:b-dju3f8) ([raw file](https://github.com/harness/harness-core/blob/f00b1225777fc1608e464ae2cdd8833e2b49e8f5/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L163)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

agree.
</blockquote></details>

This code will need refactoring anyways when I add infra step in the next PR. So I can make the changes there for both env and infra logic.

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/CustomStageConfig.java (hunk 1 above)
Comment (lovish1999) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/CustomStageConfig.java:48]: As discussed over slack, we are going ahead with this approach.

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger messagecheck

**Review by ivan-83 — DISMISSED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52719#-:-NeAJ8Gd6P5kTsMglprI:bnfp4nl)* status: 1 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by naidu-harness — APPROVED** [review]

**Comment by lovish1999** [comment]
trigger review

**Review by yogesh-chauhan — DISMISSED** [review]

**Comment by lovish1999** [comment]
trigger uts

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52719#-:-NeCtvd32g1dls4uZ-3-:bnfp4nl)* status: 4 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger uts

**Pull request merged**